lucene/core/src/java/org/apache/lucene/index/EagerCheapMergePolicy.java

```diff
@@ -0,0 +1,204 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.index;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+import org.apache.lucene.util.InfoStream;
+
+/**
+ * Eagerly proposes cheap merges on a commit. The goal is to help keep the segment count low where
+ * we can do so cheaply, particularly on a commit so that the IndexSearcher has fewer segments to
+ * search over.  This policy only overrides {@link #findFullFlushMerges(MergeTrigger, SegmentInfos, MergeContext)}; the
+ * rest are delegated to another policy.
+ */
+public class EagerCheapMergePolicy extends FilterMergePolicy {
+
+  private static final String INFO_STREAM_CATEGORY = "CHEAP_MP";
+
+  private long cheapMergeThresholdBytes = 2 * 1024 * 1024; // 2MB
+  private int cheapMinMergeAtOnce = 3;
+  private double cheapMaxSizeRatio = 1.0;
+  private int cheapLimitConcurrentMergeSegments = 0; // thus only when no merges in-progress
+
+  /**
+   * Decorates the given merge policy.
+   *
+   * @param in the wrapped {@link MergePolicy}
+   */
+  public EagerCheapMergePolicy(MergePolicy in) {
+    super(in);
+  }
+
+  /** Returns {@link #setCheapMergeThresholdMB(double)}. */
+  public long getCheapMergeThresholdBytes() {
+    return cheapMergeThresholdBytes;
+  }
+
+  /** Limit cheap merges to those less than this size in megabytes. */
+  public EagerCheapMergePolicy setCheapMergeThresholdMB(double v) {
+    v *= 1024 * 1024;
+    cheapMergeThresholdBytes = (v > Long.MAX_VALUE) ? Long.MAX_VALUE : (long) v;
+    return this;
+  }
+
+  /** Returns {@link #setCheapMinMergeAtOnce(int)}. */
+  public int getCheapMinMergeAtOnce() {
+    return cheapMinMergeAtOnce;
+  }
+
+  /** Must have at least this many cheap segments to suggest merging them. */
+  public EagerCheapMergePolicy setCheapMinMergeAtOnce(int cheapMinMergeAtOnce) {
+    this.cheapMinMergeAtOnce = cheapMinMergeAtOnce;
+    return this;
+  }
+
+  /** Returns {@link #setCheapMaxSizeRatio(double)}. */
+  public double getCheapMaxSizeRatio() {
+    return cheapMaxSizeRatio;
+  }
+
+  /**
+   * Limit the maximum ratio of size between a candidate segment (that is deemed cheap) and the sum
+   * of all smaller segments. Example: If it's 2.0, a 20KB segment will merge with a 10KB
+   * segment (or say two 5KB segments) but will not with a 9KB segment.
+   * The limit should be tuned if {@link #setCheapMinMergeAtOnce(int)} is changed. It should be more
+   * than 1/(min-1).
+   */
+  public EagerCheapMergePolicy setCheapMaxSizeRatio(double cheapMaxSizeRatio) {
+    if (cheapMaxSizeRatio <= 0.0) {
+      throw new IllegalArgumentException("must be > 0");
```

Comment:
Include the actual (invalid) value that was passed in here, in the exception message?